

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Frequently Asked Questions &mdash; MacSyLib 1.0.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
      <link rel="stylesheet" type="text/css" href="../_static/_css/custom.css?v=b89f2dfc" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=71272d9f"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Modeller Guide" href="../modeler_guide/index.html" />
    <link rel="prev" title="MacSyLib’s search engine" href="functioning.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            MacSyLib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">User Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="index.html#using-macsylib">Using MacSyLib</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#macsylib-functioning">MacSyLib functioning</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html#frequently-asked-questions">Frequently Asked Questions</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Frequently Asked Questions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#how-to-report-an-issue">How to report an issue?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-cite-macsylib-and-published-macy-models">How to cite MacSyLib and published macy-models?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-use-macsylib">How to use MacSyLib ?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-search-mode-to-be-used">What search mode to be used?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-deal-with-fragmented-genomes-mags-sags-draft-genomes">How to deal with fragmented genomes (MAGs, SAGs, draft genomes)?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#how-to-interpret-the-results-from-an-unordered-search">How to interpret the results from an <cite>unordered</cite> search?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#where-to-find-macsylib-models">Where to find MacSyLib models?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-are-the-rules-for-options-precedence">What are the rules for options precedence?</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modeler_guide/index.html">Modeller Guide</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer_guide/index.html">Developer Guide</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MacSyLib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">User Guide</a></li>
      <li class="breadcrumb-item active">Frequently Asked Questions</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="frequently-asked-questions">
<span id="faq"></span><h1>Frequently Asked Questions<a class="headerlink" href="#frequently-asked-questions" title="Link to this heading"></a></h1>
<section id="how-to-report-an-issue">
<h2>How to report an issue?<a class="headerlink" href="#how-to-report-an-issue" title="Link to this heading"></a></h2>
<p>If you encounter a problem while running MacSyLib, please submit an issue on the dedicated page of the <a class="reference external" href="https://github.com/gem-pasteur/macsylib/issues">GitHub project</a></p>
<p>To ensure we have all elements to help, please provide:</p>
<ul class="simple">
<li><p>a concise description of the issue</p></li>
<li><p>the expected behavior VS observed one</p></li>
<li><p>the exact command-line used</p></li>
<li><p>the version of MacSyLib used</p></li>
<li><p>the exact error message, and if applicable, the <cite>&lt;macsylib&gt;.log</cite> and <cite>&lt;macsylib&gt;.conf</cite> files</p></li>
<li><p>if applicable, an archive (or link to it) with the output files obtained</p></li>
<li><p>if possible, the smallest dataset there is to reproduce the issue</p></li>
<li><p>if applicable, this would also include the macsy-models (XML models plus HMM profiles) used (or precise version of the models if there are publicly available). Same as above, if possible, please provide the smallest set possible of models and HMM profiles.</p></li>
</ul>
<p>All these will definitely help us to help you! ;-)</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you use <code class="docutils literal notranslate"><span class="pre">macsylib</span></code> in higher level script you can change &lt;macsylib&gt; by the name of your tool by setting it in <code class="xref py py-class docutils literal notranslate"><span class="pre">macsylib.config.MacsyDefault</span></code> prog_name parameter.</p>
</div>
</section>
<section id="how-to-cite-macsylib-and-published-macy-models">
<span id="citations"></span><h2>How to cite MacSyLib and published macy-models?<a class="headerlink" href="#how-to-cite-macsylib-and-published-macy-models" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.24072/pcjournal.250">Néron, Bertrand et al 2023</a>, <em>Peer Community Journal</em> MacSyFinder v2: Improved modelling and search engine to identify molecular systems in genomes.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pgen.1002983">Abby and Rocha 2012</a>, <em>PLoS Genetics</em>, for the study of the evolutionary relationship between the T3SS and the bacterial flagellum, and how were designed the corresponding HMM protein profiles.</p></li>
<li><p><a class="reference external" href="https://www.nature.com/articles/srep23080">Abby et al. 2016</a>, <em>Scientific Reports</em>, for the description of bacterial protein secretion systems’ models (TXSScan: T1SS, T2SS, T5SS, T6SS, T9SS, Tad, T4P).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.pbio.3000390">Denise et al. 2019</a>, <em>PLoS Biology</em>, for the description of type IV-filament super-family models (TFF-SF: T2SS, T4aP, T4bP, Com, Tad, archaeal T4P).</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.ppat.1006525">Rendueles et al. 2017</a>, <em>PLoS Pathogens</em>, for the CapsuleFinder set of models.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1093/nar/gky425">Couvin, Bernheim et al. 2018</a>, <em>Nucleic Acids Research</em>, for the updated version of the set of Cas systems’ models, CasFinder.</p></li>
</ul>
</section>
<section id="how-to-use-macsylib">
<span id="python-example"></span><h2>How to use MacSyLib ?<a class="headerlink" href="#how-to-use-macsylib" title="Link to this heading"></a></h2>
<p>Here are an example of analyse with a python script using macsylib:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">argparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">Namespace</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.config</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.registries</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.utils</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.search_systems</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">macsylib.io</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">macsylib.system</span><span class="w"> </span><span class="kn">import</span> <span class="n">HitSystemTracker</span>

<span class="n">defaults</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">MacsyDefaults</span><span class="p">()</span>
<span class="n">settings</span> <span class="o">=</span> <span class="n">Namespace</span><span class="p">(</span>
    <span class="n">db_type</span><span class="o">=</span><span class="s1">&#39;ordered_replicon&#39;</span><span class="p">,</span>
    <span class="n">sequence_db</span><span class="o">=</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">,</span>
    <span class="n">models</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TXSScan&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">],</span>  <span class="c1"># this model must be installed with msl_data scripts</span>
    <span class="n">worker</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
    <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;my_results&#39;</span>
<span class="p">)</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">Config</span><span class="p">(</span><span class="n">defaults</span><span class="p">,</span> <span class="n">settings</span><span class="p">)</span>

<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">working_dir</span><span class="p">())</span>  <span class="c1"># working_dir = out_dir; you have to create this directory</span>

<span class="n">macsylib</span><span class="o">.</span><span class="n">init_logger</span><span class="p">(</span><span class="n">log_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">working_dir</span><span class="p">(),</span> <span class="n">config</span><span class="o">.</span><span class="n">log_file</span><span class="p">()))</span>
<span class="n">macsylib</span><span class="o">.</span><span class="n">logger_set_level</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;macsylib&#39;</span><span class="p">)</span>
<span class="n">model_registry</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">registries</span><span class="o">.</span><span class="n">ModelRegistry</span><span class="p">()</span>

<span class="k">for</span> <span class="n">model_dir</span> <span class="ow">in</span> <span class="n">config</span><span class="o">.</span><span class="n">models_dir</span><span class="p">():</span>
    <span class="n">models_loc_available</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">registries</span><span class="o">.</span><span class="n">scan_models_dir</span><span class="p">(</span><span class="n">model_dir</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">model_loc</span> <span class="ow">in</span> <span class="n">models_loc_available</span><span class="p">:</span>
        <span class="n">model_registry</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">model_loc</span><span class="p">)</span>
<span class="n">models_def_to_detect</span><span class="p">,</span> <span class="n">models_fam_name</span><span class="p">,</span> <span class="n">models_version</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">get_def_to_detect</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">models</span><span class="p">(),</span>
                                                                                         <span class="n">model_registry</span><span class="p">)</span>

<span class="n">all_systems</span><span class="p">,</span> <span class="n">rejected_candidates</span> <span class="o">=</span> <span class="n">macsylib</span><span class="o">.</span><span class="n">search_systems</span><span class="o">.</span><span class="n">search_systems</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">model_registry</span><span class="p">,</span> <span class="n">models_def_to_detect</span><span class="p">,</span>
                                                                          <span class="n">logger</span><span class="p">)</span>
<span class="n">track_multi_systems_hit</span> <span class="o">=</span> <span class="n">HitSystemTracker</span><span class="p">(</span><span class="n">all_systems</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">working_dir</span><span class="p">(),</span><span class="s1">&#39;all_systems.txt&#39;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tsv_file</span><span class="p">:</span>
    <span class="n">macsylib</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">systems_to_tsv</span><span class="p">(</span><span class="s1">&#39;TXSScan&#39;</span><span class="p">,</span> <span class="s1">&#39;1.1.3&#39;</span><span class="p">,</span>
                               <span class="n">all_systems</span><span class="p">,</span>
                               <span class="n">track_multi_systems_hit</span><span class="p">,</span>
                               <span class="n">tsv_file</span><span class="p">,</span>
                               <span class="n">header</span><span class="o">=</span><span class="k">lambda</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">model_v</span><span class="p">,</span> <span class="n">skipped_replicons</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;# created by </span><span class="si">{</span><span class="vm">__file__</span><span class="si">}</span><span class="s1"> script with models </span><span class="si">{</span><span class="n">model_name</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">model_v</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The script above will produce the file below with a summary of all systems found.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># created by /home/bneron/Projects/GEM/MacSyFinder/src/macsylib/Sandbox/msl_example.py script with models TXSScan-1.1.3
# Systems found:
replicon	hit_id	gene_name	hit_pos	model_fqn	sys_id	sys_loci	locus_num	sys_wholeness	sys_score	sys_occ	hit_gene_ref	hit_status	hit_seq_len	hit_i_eval	hit_score	hit_profile_cov	hit_seq_cov	hit_begin_match	hit_end_match	counterpart	used_in
test	VICH001.B.00001.C001_01397	T1SS_abc	11	TXSScan/bacteria/diderm/T1SS	test_T1SS_1	1	1	1.000	2.700	2	T1SS_abc	mandatory	721	2.4e-157	516.300	0.996	0.698	175	677		
test	VICH001.B.00001.C001_01398	T1SS_mfp	12	TXSScan/bacteria/diderm/T1SS	test_T1SS_1	1	1	1.000	2.700	2	T1SS_mfp	mandatory	467	2.3e-72	235.600	1.000	0.797	85	456		
test	VICH001.B.00001.C001_01399	T1SS_abc	13	TXSScan/bacteria/diderm/T1SS	test_T1SS_1	1	1	1.000	2.700	2	T1SS_abc	mandatory	720	2.1e-158	519.700	1.000	0.699	167	669		
test	VICH001.B.00001.C001_01506	T1SS_omf	23	TXSScan/bacteria/diderm/T1SS	test_T1SS_1	1	-1	1.000	2.700	2	T1SS_omf	mandatory	419	9.3e-35	111.500	0.998	0.912	25	406		

test	VICH001.B.00001.C001_01397	T1SS_abc	11	TXSScan/bacteria/diderm/T1SS	test_T1SS_2	1	1	1.000	2.700	2	T1SS_abc	mandatory	721	2.4e-157	516.300	0.996	0.698	175	677		
test	VICH001.B.00001.C001_01398	T1SS_mfp	12	TXSScan/bacteria/diderm/T1SS	test_T1SS_2	1	1	1.000	2.700	2	T1SS_mfp	mandatory	467	2.3e-72	235.600	1.000	0.797	85	456		
test	VICH001.B.00001.C001_01399	T1SS_abc	13	TXSScan/bacteria/diderm/T1SS	test_T1SS_2	1	1	1.000	2.700	2	T1SS_abc	mandatory	720	2.1e-158	519.700	1.000	0.699	167	669		
test	VICH001.B.00001.C001_01506	T1SS_omf	23	TXSScan/bacteria/diderm/T1SS	test_T1SS_2	1	-1	1.000	2.700	2	T1SS_omf	mandatory	419	9.3e-35	111.500	0.998	0.912	25	406		
</pre></div>
</div>
<p>The code and the data are available</p>
<p><a class="reference download internal" download="" href="../_downloads/5b3c3c607624b82117b181b8821d85d3/msl_example.py"><code class="xref download docutils literal notranslate"><span class="pre">msl_example.py</span></code></a> .</p>
<p><a class="reference download internal" download="" href="../_downloads/1b495793e7eacfaf5617fda19eba0683/test.fasta"><code class="xref download docutils literal notranslate"><span class="pre">test.fasta</span></code></a> .</p>
<div class="line-block">
<div class="line">For more details check the developer guide <a class="reference internal" href="../developer_guide/index.html#developer-guide"><span class="std std-ref">Developer Guide</span></a> and api documentation <a class="reference internal" href="../developer_guide/api/index.html#api"><span class="std std-ref">MacSyLib API documentation</span></a></div>
<div class="line">For more example check <a class="reference external" href="https://github.com/gem-pasteur/macsyfinder/tree/master">macsyfinder source code</a></div>
</div>
</section>
<section id="what-search-mode-to-be-used">
<span id="faq-search-mode"></span><h2>What search mode to be used?<a class="headerlink" href="#what-search-mode-to-be-used" title="Link to this heading"></a></h2>
<p>Depending on the type of dataset you have, you will have to adapt MacSyLib’s search mode.</p>
<ul class="simple">
<li><p>If you have a fasta file from a complete genome where <strong>proteins are ordered</strong> according to the corresponding genes’ order along the replicon,
your dataset is entitled to the most powerful search mode (see below): <cite>ordered_replicon</cite> and use the following option <cite>–db-type ordered_replicon</cite>.</p></li>
<li><p>If you have a fasta file of proteins with <strong>no sense of the order</strong> of the corresponding genes along the chromosome(s) or replicon(s),
you will have to use the <cite>unordered</cite> search mode with the following option: <cite>–db-type unordered</cite></p></li>
<li><p>If you have <strong>multiple ordered replicons</strong> to analyse at once, you can follow the <cite>Gembase</cite> convention to name the proteins in the fasta file,
so that the original replicons can be assessed from their name: <a class="reference internal" href="gembase_convention.html#gembase-convention"><span class="std std-ref">see here for a description</span></a>.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>When the <strong>gene order is known</strong> (<cite>ordered_replicon</cite> search mode) the power of the analysis is <strong>maximal</strong>,
since both the genomic content and context are taken into account for the search.</p></li>
<li><p>When the <strong>gene order is unknown</strong> (<cite>unordered</cite> search mode) the power of the analysis is more <strong>limited</strong> since
the presence of systems can only be suggested on the basis of the quorum of components - and not based on genomic context information.</p></li>
</ul>
</div>
<p>More on MacSyLib’s functioning <a class="reference internal" href="functioning.html#functioning"><span class="std std-ref">here</span></a>.</p>
</section>
<section id="how-to-deal-with-fragmented-genomes-mags-sags-draft-genomes">
<h2>How to deal with fragmented genomes (MAGs, SAGs, draft genomes)?<a class="headerlink" href="#how-to-deal-with-fragmented-genomes-mags-sags-draft-genomes" title="Link to this heading"></a></h2>
<p>There are more and more genomes available which are not completely assembled, or are fragmented and incomplete.
In this case, several options can be considered.</p>
<p>1. If your genome is at least partially assembled and contigs are not too short, you might “feel lucky” and first
consider to run MacSyLib with the <cite>ordered_replicon</cite> mode. It could be particularly efficient if you are investigating
systems encoded by compact loci (Cas systems, some secretion systems…), as they might be encoded by a single contig.</p>
<p>2. On top of the <cite>ordered_replicon</cite> mode, you might add the option “multi-loci” to the systems to annotate (if not already the case),
in order to maximize the chance to annotate an entire system, even if encoded across several contigs.</p>
<p>3. The <cite>unordered</cite> mode can be used in complement of the two above options, e.g. to retrieve some of the missing components.
It will enable to assess the genetic potential and possible presence of a system, independently of the quality of assembly of the genome.
It might also be the only reasonable option if the genome is too fragmented and/or too incomplete.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The results obtained with the <cite>ordered_replicon</cite> mode on a fragmented genome have to be considered carefully, especially with respect
to the contigs’ borders, as some proteins from different contigs might be artificially considered as closely encoded.</p></li>
<li><p>To retrieve “fragments” of a system not found to reach the quorum in the <cite>ordered_replicon</cite> mode, it is possible to retrieve
clusters of genes from the <em>rejected_candidates.tsv</em> file.</p></li>
</ul>
</div>
</section>
<section id="how-to-interpret-the-results-from-an-unordered-search">
<h2>How to interpret the results from an <cite>unordered</cite> search?<a class="headerlink" href="#how-to-interpret-the-results-from-an-unordered-search" title="Link to this heading"></a></h2>
<p>As mentioned above, in the <cite>unordered</cite> search mode, the inference of a system’s presence is only based on the list of components found in the protein dataset.
Thus, the kind of search specificity provided when using the genomic context
(components next to each other are more likely to be part of a same system’s occurrence) is not within reach.</p>
<p>In the <cite>unordered</cite> search mode, the number of proteins selected as system’s components
(based on the filtering of HMM profiles’ similarity search) is reported.
We decided to report all kinds of system’s components, including the <cite>forbidden</cite> ones in order to raise awareness of
the user -&gt; even if all constraints are met for the system’s inference
(here, the quorum: minimal number of components), it cannot be excluded that a <cite>forbidden</cite>
component would lie next to the <em>bona fide</em> components (<cite>mandatory</cite> and <cite>accessory</cite> ones) in the genome…</p>
<p>In the end, the <cite>unordered</cite> search mode provides an idea as to whether the <strong>genetic potential</strong> for
a given system is found in the set of proteins analysed, with no attempt to assign proteins to particular systems’ occurrences,
nor guarantee as to whether <cite>forbidden</cite> components should be considered for the potential occurrences.</p>
</section>
<section id="where-to-find-macsylib-models">
<h2>Where to find MacSyLib models?<a class="headerlink" href="#where-to-find-macsylib-models" title="Link to this heading"></a></h2>
<p>Since version 2, there is a tool to enable the download and installation of published models from a repository: the <cite>msl_data</cite> tool.</p>
<p>See <a class="reference internal" href="installation.html#msl-data"><span class="std std-ref">here for details</span></a> on how to use it.</p>
</section>
<section id="what-are-the-rules-for-options-precedence">
<h2>What are the rules for options precedence?<a class="headerlink" href="#what-are-the-rules-for-options-precedence" title="Link to this heading"></a></h2>
<p>MacSyLib offers many ways to parametrize the systems’ search: through the command-line,
through various configuration files (for the models, for the run, etc…).
It offers a large control over the search engine. But it also means you can get lost in configuration. ;-)</p>
<p>Here is a recap of the rules for options precedence. In a general manner, the command line always wins.</p>
<p>The precedence rules between the different levels of configuration are:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>system &lt; home &lt; model &lt; project &lt; --cfg-file | --previous-run &lt; command line options
</pre></div>
</div>
<ul class="simple">
<li><p><strong>system</strong>: the <cite>&lt;macsylib&gt;.conf</cite> file either in ${VIRTUAL_ENV}/etc/&lt;macsylib&gt;/
in case of a <em>virtualenv</em> this configuration affects only the MacSyLib version installed in this virtualenv</p></li>
<li><p><strong>home</strong>:  the <cite>~/.&lt;macsylib&gt;/&lt;macsylib&gt;.conf</cite> file</p></li>
<li><p><strong>model</strong>: the <cite>model_conf.xml</cite> file at the root of the model package</p></li>
<li><p><strong>project</strong>: the <cite>&lt;macsylib&gt;.conf</cite> file found in the directory where the <cite>macsylib</cite> command was run</p></li>
<li><p><strong>cfgfile</strong>: any configuration file specified by the user on the command line</p></li>
<li><p><strong>previous-run</strong>: the <cite>&lt;macsylib&gt;.conf</cite> file found in the results directory of the previous run</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="functioning.html" class="btn btn-neutral float-left" title="MacSyLib’s search engine" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../modeler_guide/index.html" class="btn btn-neutral float-right" title="Modeller Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2014-2025, Bertrand Néron, Sophie Abby.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
  

<div class="licensing">
  <p>
    This documentation is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
      <img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png">.
    </a>
  </p>
</div>


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

<a class="github" href="https://github.com/gem-pasteur/macsylib">
    <img style="position: absolute; top: 0; right: 0; border: 0;"
         src="https://raw.githubusercontent.com/gem-pasteur/macsylib/refs/heads/main/doc/source/_static/github-banner.svg"
         alt="Fork me on GitHub" />
</a>



</body>
</html>